File: src/_libs/search/duckduckgo.ts
import { load } from 'cheerio'
import type { SearchResponse } from './types'

export async function searchDuckDuckGo(query: string): Promise<SearchResponse> {
  const url = `https://html.duckduckgo.com/html/?q=${encodeURIComponent(query)}`
  const res = await fetch(url)
  const html = await res.text()
  const $ = load(html)

  const results = $('.result')
    .map((_, item) => {
      const title = $(item).find('.result__a').text().trim()
      const url = $(item).find('.result__url').text().trim()
      const snippet = $(item).find('.result__snippet').text().trim()
      if (title && url) return { title, url, snippet }
    })
    .get()

  let prev: string | undefined
  let next: string | undefined

  const navLinks = $('.nav-link a')
  navLinks.each((_, link) => {
    const text = $(link).text().trim()
    const href = $(link).attr('href')
    if (href) {
      if (/prev/i.test(text)) prev = href
      if (/next/i.test(text)) next = href
    }
  })

  const forms = $('.nav-link form')
  forms.each((_, form) => {
    const value = $(form).find('input[type="submit"]').attr('value') ?? ''
    const action = $(form).attr('action') ?? '/html/'
    const params = new URLSearchParams()
    $(form)
      .find('input[type="hidden"]')
      .each((_, input) => {
        params.set($(input).attr('name') ?? '', $(input).attr('value') ?? '')
      })
    const formUrl = `https://html.duckduckgo.com${action}?${params.toString()}`
    if (/prev/i.test(value)) prev = formUrl
    if (/next/i.test(value)) next = formUrl
  })

  return { results, prev, next }
}

File: internal/lib/search/duckduckgo.go
package search

import (
	"fmt"
	"net/url"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/enetx/g"
	"github.com/enetx/surf"
)

func DuckDuckGo(query string) (*Response, error) {
	searchURL := "https://html.duckduckgo.com/html/?q=" + url.QueryEscape(query)

	client := surf.NewClient()
	result := client.Get(g.NewString(searchURL)).Do()
	if result.IsErr() {
		return nil, fmt.Errorf("fetch duckduckgo: %w", result.Err())
	}
	resp := result.Ok()
	defer resp.Body.Close()

	doc, err := goquery.NewDocumentFromReader(resp.Body.Reader)
	if err != nil {
		return nil, fmt.Errorf("parse duckduckgo: %w", err)
	}

	var results []Result
	doc.Find(".result").Each(func(i int, s *goquery.Selection) {
		title := strings.TrimSpace(s.Find(".result__a").Text())
		resultURL := strings.TrimSpace(s.Find(".result__url").Text())
		snippet := strings.TrimSpace(s.Find(".result__snippet").Text())
		if title != "" && resultURL != "" {
			results = append(results, Result{
				Title:   title,
				URL:     resultURL,
				Snippet: snippet,
			})
		}
	})

	var prev, next string

	doc.Find(".nav-link a").Each(func(i int, s *goquery.Selection) {
		text := strings.TrimSpace(s.Text())
		href, exists := s.Attr("href")
		if exists {
			if strings.Contains(strings.ToLower(text), "prev") {
				prev = href
			}
			if strings.Contains(strings.ToLower(text), "next") {
				next = href
			}
		}
	})

	doc.Find(".nav-link form").Each(func(i int, s *goquery.Selection) {
		value := strings.TrimSpace(s.Find("input[type=\"submit\"]").AttrOr("value", ""))
		action := s.AttrOr("action", "/html/")

		params := url.Values{}
		s.Find("input[type=\"hidden\"]").Each(func(j int, input *goquery.Selection) {
			name := input.AttrOr("name", "")
			val := input.AttrOr("value", "")
			if name != "" {
				params.Set(name, val)
			}
		})

		formURL := fmt.Sprintf("https://html.duckduckgo.com%s?%s", action, params.Encode())
		if strings.Contains(strings.ToLower(value), "prev") {
			prev = formURL
		}
		if strings.Contains(strings.ToLower(value), "next") {
			next = formURL
		}
	})

	return &Response{
		Results: results,
		Prev:    prev,
		Next:    next,
	}, nil
}
